core-bundle/src/Util/SimpleTokenParser.php

```diff
@@ -0,0 +1,213 @@
+                    if (null !== $this->logger) {
+                        $this->logger->log(LogLevel::INFO, sprintf('Tried to parse unknown simple token "%s".', $matches[1]));
+                    }
+
+                    return '##'.$matches[1].'##';
+                }
+
+                return $data[$matches[1]];
+            },
+            $subject
+        );
+    }
+
+    private function canUseExpressionLanguage(array $data): bool
+    {
+        foreach (array_keys($data) as $token) {
+            if (!preg_match('/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/', $token)) {
+                @trigger_error('Using tokens that are not valid PHP variables has been deprecated and will no longer work in Contao 5.0. Falling back to legacy token parsing.', E_USER_DEPRECATED);
+
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function evaluateExpression(string $expression, array $data, bool $canUseExpressionLanguage): bool
+    {
+        if ($canUseExpressionLanguage) {
+            try {
+                return $this->expressionLanguage->evaluate($expression, $data);
+            } catch (\Exception $e) {
+                throw new \InvalidArgumentException($e->getMessage(), 0, $e);
+            }
+        }
+
+        // Legacy code
+        if (!preg_match('/^([^=!<>\s]+) *([=!<>]+)(.+)$/s', $expression, $matches)) {
+            return false;
+        }
+
+        [, $token, $operator, $value] = $matches;
+
+        if (!\array_key_exists($token, $data)) {
+            if (null !== $this->logger) {
+                $this->logger->log(LogLevel::INFO, sprintf('Tried to evaluate unknown simple token "%s".', $token));
+            }
+
+            return false;
+        }
+
+        $tokenValue = $data[$token];
+
+        // Normalize types
+        $value = trim($value, ' ');
+
+        if (is_numeric($value)) {
+            if (false === strpos($value, '.')) {
+                $value = (int) $value;
+            } else {
+                $value = (float) $value;
+            }
+        } elseif ('true' === strtolower($value)) {
+            $value = true;
+        } elseif ('false' === strtolower($value)) {
+            $value = false;
+        } elseif ('null' === strtolower($value)) {
+            $value = null;
+        } elseif (0 === strncmp($value, '"', 1) && '"' === substr($value, -1)) {
+            $value = str_replace('\"', '"', substr($value, 1, -1));
+        } elseif (0 === strncmp($value, "'", 1) && "'" === substr($value, -1)) {
+            $value = str_replace("\\'", "'", substr($value, 1, -1));
+        } else {
+            throw new \InvalidArgumentException(sprintf('Unknown data type of comparison value "%s".', $value));
+        }
+
+        // Evaluate
+        switch ($operator) {
+            case '==':
+                // todo: find a way to stop the CS fixer to change this to ===
+                return $tokenValue == $value;
```

Comment:
```suggestion
                return \in_array($tokenValue, [$value]);
```
This would be not allowed too by the CS fixer i guess?

Reply:
The CS fixer will probably change it to `\in_array($tokenValue, [$value], true)`, which is the same as `$tokenValue === $value`.

Reply:
```suggestion
                switch ($tokenValue) {
                    case $value:
                        return true;
                    default:
                        return false;
                }
```

Reply:
If I explicitly set the third parameter to false it works. :smile:

Reply:
Nice find!

Reply:
It can't be optimized by opcode anymore though :) So might as well kill the global function namespace but not sure if the fixer accepts that then :D

Reply:
The `\` prefix gets added automatically.